### PR TITLE
[PIP 101] Add seek by index feature for consumer

### PR DESCRIPTION
**Pull request by Jason918**
### Motivation

Currently we can reset the read position of a cursor by message id or timestamp. Since we formerly introduced index in broker metadata since 2.8.0, reset cursor by index is very helpful in other protocol handler (KoP or RoP).

### Modifications

see #12234 

### Verifying this change

- [x] Make sure that the change passes the CI checks.

This change added tests and can be verified as follows:
  - PersistentMessageFinderByIndexTest.java is unit test for the core finder implementation in broker.
  - ConsumerSeekTest.java is the integration test for client implementation.

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): ( no)
  - The public API: (yes?)  added optional field 'index' in CommandSeek
  - The schema: ( no ) 
  - The default values of configurations: (no)
  - The wire protocol: (no)
  - The rest endpoints: (no)
  - The admin cli options: (no)
  - Anything that affects deployment: ( no)

### Documentation

Check the box below and label this PR (if you have committer privilege).
  
- [x] doc 
  
  (If this PR contains doc changes) 
User can refer to the java doc in file pulsar-client-api/src/main/java/org/apache/pulsar/client/api/Consumer.java

### CODE REVIEW

**Comment by MarvinCai** [comment]
it looks like we try to bake some sort of support for other message protocol into the broker itself?
will this just open up the potential to add more patches to make supporting other message protocol easier, then how far are we willing to go?

**Comment by Jason918** [comment]
> it looks like we try to bake some sort of support for other message protocol into the broker itself?
> will this just open up the potential to add more patches to make supporting other message protocol easier, then how far are we willing to go?

Message protocol handler is one of the great features in pulsar.
IMHO, It's important and useful to make other message protocol implementation easier. It's great for the messaging ecosystem. So we should add more new features supporting other message protocol as long as it does not breaking the design philosophy of pulsar.

@codelipenghui @jiazhai  What do you think of this?

**Comment by BewareMyPower** [comment]
@MarvinCai It's not a support for other protocol, it's an enhancement. Currently Pulsar support the seek operation based on timestamp or message id. However, it doesn't support seek operation based on message count. For example, users might want to seek to 10000 messages before. Currently they cannot achieve this goal.
Pulsar 2.8.0 introduced BrokerEntryMetadata (PIP-70) and the message index (similar to Kafka's offset), so it becomes possible to support seeking by index now.

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ConsumerSeekTest.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    @AfterClass
+    @Override
+    protected void cleanup() throws Exception {
+        internalCleanup();
+    }
+
+    @Test
+    public void testConsumerSeekByIndex() throws Exception {
```
Comment (BewareMyPower) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ConsumerSeekTest.java:59]: It's better to move the test to `SubscriptionSeekTest`. And I don't think validating `markDeletePosition` is a good way, you can take other tests in `SubscriptionSeekTest` as reference.

File: pulsar-client-api/src/main/java/org/apache/pulsar/client/api/Consumer.java [hunk 2]
```diff
@@ -599,6 +599,15 @@
      */
     void seek(long timestamp) throws PulsarClientException;
 
+
+    /**
+     * Reset the subscription associated with this consumer to a specific message index.
+     *
+     * @param index
+     *            the message index where to reposition the subscription
+     */
```
Comment (BewareMyPower) [inline pulsar-client-api/src/main/java/org/apache/pulsar/client/api/Consumer.java:608]: We should define the behavior for corner cases. For example, if there're 10 messages in a topic, i.e. the offset range varies from 0 to 9. What will happen if the given index is -1 or 10?

There is another corner case. If the offset range is `[10, 20)`, the oldest messages (`[0, 10)`) have been deleted. What will happen if the given index is less than 10?

BTW, it's a change for public API. I think we need a PIP first. /cc @merlimat

**Review by BewareMyPower — CHANGES_REQUESTED** [review]

**Comment by MarvinCai** [comment]
But not all users will use Index or event BrokerEntryMetadata, while all messages have Timestamp and MessageID, that's a bit different in my opinion. Adding a method that rely on a property might not exist for all users is kind of weird.

**Comment by MarvinCai** [comment]
But let's discuss it with the community.

**Comment by Jason918** [comment]
Closing this for now, as there are not enough votes.

**Pull request closed without merging**